# Cleaning and Analyzing a dataset of used cars from the German eBay website

* Dataset contains ~50,000 data points. The variables described by the dataset are:
    - dateCrawled - When this ad was first crawled. All field-values are taken from this date.
    - name - Name of the car.
    - seller - Whether the seller is private or a dealer.
    - offerType - The type of listing
    - price - The price on the ad to sell the car.
    - abtest - Whether the listing is included in an A/B test.
    - vehicleType - The vehicle Type.
    - yearOfRegistration - The year in which the car was first registered.
    - gearbox - The transmission type.
    - powerPS - The power of the car in PS.
    - model - The car model name.
    - kilometer - How many kilometers the car has driven.
    - monthOfRegistration - The month in which the car was - first registered.
    - fuelType - What type of fuel the car uses.
    - brand - The brand of the car.
    - notRepairedDamage - If the car has a damage which is not yet repaired.
    - dateCreated - The date on which the eBay listing was created.
    - nrOfPictures - The number of pictures in the ad.
    - postalCode - The postal code for the location of the vehicle.
    - lastSeenOnline - When the crawler saw this ad last online.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [3]:
autos.head(10)

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6  2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7  2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8  2016-03-22 16:51:34                                         Seat_Arosa   
9  2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   
6  privat   Angebot    $300     test   limousine                1995   
7  privat   Angebot  $1,990  control   limousine                1998   
8  privat   Angebot    $250     test         NaN                2000   
9  privat   Angebot    $590  control         bus                1997   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   
6    manuell       90     golf  150,000km                    8   benzin   
7    manuell       90     golf  150,000km                   12   diesel   
8    manuell        0    arosa  150,000km                   10      NaN   
9    manuell       90   megane  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   
6  volkswagen               NaN  2016-03-20 00:00:00             0   
7  volkswagen              nein  2016-03-16 00:00:00             0   
8        seat              nein  2016-03-22 00:00:00             0   
9     renault              nein  2016-03-16 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21  
6       31535  2016-03-23 02:48:59  
7       53474  2016-04-07 03:17:32  
8        7426  2016-03-26 18:18:10  
9       15749  2016-04-06 10:46:35

Initial inspection of the dataset reveals that 5 variables having missing values ranging from ~3700 to ~10000 data points per variable. 

These need to addressed in an appropriate manner before proceeding with the analysis

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

In [6]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

Column names were changed from Camelcase to Snakecase. 

Some column names were changed to improve ease of understanding.

In [7]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-05 16:57:05  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

## Initial diagnosis of the dataset

The following issues need to addressed before we can proceed with the analysis.

- 3 columns ('seller', 'offer_type', and 'nr_of_pictures') do not seem to add much value to the analysis. 


- 'seller' and 'offer_type' contain only 2 unique types of observations and are dominated by 1 type in each case. 


- None of the rows have any pictures and hence the 'nr_of_pictures' is practically a null column and can be excluded safely.


- 'price' and 'odometer' are stored as strings and need to converted to int64 datatype.


- 'registration_year' needs cleaning since the minimum recorded year is 1000 and the maximum is 9999, both of which are nonsensical.


- 'registration_month' has a minimum value of 0 which does not make sense.


- 'power_ps' has a maximum value that is more than 100 times the mean, needs to checked for possible outliers.

In [8]:
autos = autos.drop(["seller","offer_type","nr_of_pictures"],axis=1)

In [9]:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","").astype(int)

autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","").astype(int)

autos.rename(columns={"price":"price_$"},inplace=True)
autos.rename(columns={"odometer":"odometer_km"},inplace=True)

In [10]:
autos.describe(include = 'all')

date_crawled         name       price_$ ab_test vehicle_type  \
count                 50000        50000  5.000000e+04   50000        44905   
unique                48213        38754           NaN       2            8   
top     2016-03-05 16:57:05  Ford_Fiesta           NaN    test    limousine   
freq                      3           78           NaN   25756        12859   
mean                    NaN          NaN  9.840044e+03     NaN          NaN   
std                     NaN          NaN  4.811044e+05     NaN          NaN   
min                     NaN          NaN  0.000000e+00     NaN          NaN   
25%                     NaN          NaN  1.100000e+03     NaN          NaN   
50%                     NaN          NaN  2.950000e+03     NaN          NaN   
75%                     NaN          NaN  7.200000e+03     NaN          NaN   
max                     NaN          NaN  1.000000e+08     NaN          NaN   

        registration_year  gearbox      power_ps  model    odometer_km  \
count        50000.000000    47320  50000.000000  47242   50000.000000   
unique                NaN        2           NaN    245            NaN   
top                   NaN  manuell           NaN   golf            NaN   
freq                  NaN    36993           NaN   4024            NaN   
mean          2005.073280      NaN    116.355920    NaN  125732.700000   
std            105.712813      NaN    209.216627    NaN   40042.211706   
min           1000.000000      NaN      0.000000    NaN    5000.000000   
25%           1999.000000      NaN     70.000000    NaN  125000.000000   
50%           2003.000000      NaN    105.000000    NaN  150000.000000   
75%           2008.000000      NaN    150.000000    NaN  150000.000000   
max           9999.000000      NaN  17700.000000    NaN  150000.000000   

        registration_month fuel_type       brand unrepaired_damage  \
count         50000.000000     45518       50000             40171   
unique                 NaN         7          40                 2   
top                    NaN    benzin  volkswagen              nein   
freq                   NaN     30107       10687             35232   
mean              5.723360       NaN         NaN               NaN   
std               3.711984       NaN         NaN               NaN   
min               0.000000       NaN         NaN               NaN   
25%               3.000000       NaN         NaN               NaN   
50%               6.000000       NaN         NaN               NaN   
75%               9.000000       NaN         NaN               NaN   
max              12.000000       NaN         NaN               NaN   

                 ad_created   postal_code            last_seen  
count                 50000  50000.000000                50000  
unique                   76           NaN                39481  
top     2016-04-03 00:00:00           NaN  2016-04-07 06:17:27  
freq                   1946           NaN                    8  
mean                    NaN  50813.627300                  NaN  
std                     NaN  25779.747957                  NaN  
min                     NaN   1067.000000                  NaN  
25%                     NaN  30451.000000                  NaN  
50%                     NaN  49577.000000                  NaN  
75%                     NaN  71540.000000                  NaN  
max                     NaN  99998.000000                  NaN

In [11]:
autos_price_shape = autos["price_$"].unique().shape
print(autos_price_shape)
print("\nPrice column description:")
print(autos["price_$"].describe())
autos_price_counts = autos["price_$"].value_counts()
print("\nSorted Price value counts at the top and bottom:")
print( autos_price_counts.sort_index(ascending=True).head(20),"\n")
print( autos_price_counts.sort_index(ascending=False).head(20),"\n")

(2357,)

Price column description:
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64

Sorted Price value counts at the top and bottom:
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price_$, dtype: int64 

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price_$, dtype: int64 



We can observe from the descriptive statistics that there appear some outliers within the 'price_$' column:

*    There are 1421 rows with a price of \$0, these can be exlcuded as they only account for ~3\% of the total no. of rows.


*    There are 156 rows with a price of 1, this is common within eBay auctions where the opening bid starts at \$1.


*    When looking at the maximum prices of used cars, there appears to steep increase from car prices from \$350000 to \$999990. Since there are only a handful of cars with prices above \$350000, we could exclude them as outliers. A plausible explanation for their price is that they might be rare vintage vehicles.

In [12]:
#dropping outliers from 'price_$' column
autos = autos[autos["price_$"].between(1,350000)]

print(autos["price_$"].describe())

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price_$, dtype: float64


In [13]:
autos_odometer_shape = autos["odometer_km"].unique().shape
print(autos_odometer_shape)
print("\nOdometer column description:")
print(autos["odometer_km"].describe())
autos_odometer_counts = autos["odometer_km"].value_counts()
print("\nSorted Odometer distance counts at the top and bottom:")
print(autos_odometer_counts.sort_index(ascending=True).head(10),"\n")
print(autos_odometer_counts.sort_index(ascending=False).head(10),"\n")

(13,)

Odometer column description:
count     48565.000000
mean     125770.101925
std       39788.636804
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

Sorted Odometer distance counts at the top and bottom:
5000      836
10000     253
20000     762
30000     780
40000     815
50000    1012
60000    1155
70000    1217
80000    1415
90000    1734
Name: odometer_km, dtype: int64 

150000    31414
125000     5057
100000     2115
90000      1734
80000      1415
70000      1217
60000      1155
50000      1012
40000       815
30000       780
Name: odometer_km, dtype: int64 



The numbers do not seem to be too out of place. There are several cars with a high mileage (150000 km) on them.

In [14]:
autos_dates_crawled = autos['date_crawled'].str[:10]
autos_ad_created = autos['ad_created'].str[:10]
autos_last_seen = autos['last_seen'].str[:10]

In [15]:
print(autos_dates_crawled.value_counts(normalize=True,dropna=False).sort_index(ascending=True))

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64


There appears to be an equal distribution amongst the dates crawled. All the data appears to be crawled in 32-33 days between March 5,2016 and April 7, 2016

In [16]:
print(autos_ad_created.value_counts(normalize=True,dropna=False).sort_index(ascending=True))


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

There were few ads created in 2015 and the number steadily progressed. There was an even and significant spike in the number of ads created between February 2016 and early March 2016.

From March 2016 onwards there appears to even distribution in the number of ads created per day. This coincides with the range of dates crawled, indicating that new cars were being listed at a somewhat consistent rate during those dates.

In [17]:
print(autos_last_seen.value_counts(normalize=True,dropna=False).sort_index(ascending=True))


2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64


~55% of the ads were last seen within the first 7 days of April 2016. The last seen date of ad can be inferred as the date an ad is no longer listed on eBay. Since there is a uniform distrbution of ads last seen prior to the last 3 days in the dataset, we can conclude that the ads were delisted either due to sales or they removed due to lack of any attention from buyers. 

In [18]:
print(autos["registration_year"].describe())

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64


As we pointed out earlier, the 'registration_year' column has some anomalies. The minimum is 1000 which is not feasible as cars were created much later, while the max being 9999 is also absurd. 

Based on the last seen of the ads, we can see that we can exclude any cars that have registration dates after 2016.

In [19]:
print(autos["registration_year"].value_counts(normalize=True,dropna=False).sort_index(ascending=True))
print('\n')
print(autos["registration_year"]
      .between(1910,2016)
      .value_counts(normalize=True,dropna=False)
      .sort_index(ascending=True))

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000103
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000041
1937    0.000082
1938    0.000021
1939    0.000021
1941    0.000041
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000041
1952    0.000021
1953    0.000021
1954    0.000041
1955    0.000041
1956    0.000082
1957    0.000041
1958    0.000082
1959    0.000124
1960    0.000474
1961    0.000124
1962    0.000082
1963    0.000165
1964    0.000247
          ...   
2000    0.064985
2001    0.054278
2002    0.051189
2003    0.055575
2004    0.055657
2005    0.060455
2006    0.054978
2007    0.046886
2008    0.045609
2009    0.042932
2010    0.032719
2011    0.033419
2012    0.026974
2013    0.016535
2014    0.013652
2015    0.008072
2016    0.025121
2017    0.028663
2018    0.009678
2019    0.000041
2800    0.000021
4100    0.000021
4500    0.000021
4800    0.000021
5000    0.000082
5911    0.000021
6200    0.000021
8888    0.0000

96% of the dataset has car registered between 1910 and 2016, which is feasible since the earliest car was built is the late 1800's. Since only 4% of the dataset is outside the range we are considering we can exclude those dates.

In [20]:
autos = autos[autos["registration_year"].between(1910,2016)]
print(autos['registration_year'].describe())
print('\n')

(autos['registration_year']
    .value_counts(normalize = True))

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64




2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
2008    0.047450
2009    0.044665
1997    0.041794
2011    0.034768
2010    0.034040
1996    0.029412
2012    0.028063
1995    0.026285
2016    0.026135
2013    0.017202
2014    0.014203
1994    0.013474
1993    0.009104
2015    0.008397
1992    0.007926
1990    0.007433
1991    0.007262
1989    0.003727
1988    0.002892
1985    0.002035
          ...   
1966    0.000471
1976    0.000450
1969    0.000407
1975    0.000386
1965    0.000364
1964    0.000257
1963    0.000171
1959    0.000129
1961    0.000129
1910    0.000107
1956    0.000086
1958    0.000086
1937    0.000086
1962    0.000086
1950    0.000064
1954    0.000043
1941    0.000043
1951    0.000043
1934    0.000043
1957    0.000043
1955    0.000043
1953    0.000021
1943    0.000021
1929    0.000021
1939    0.000021
1938    0.000021
1948    0.000021
1927    0.0000

Based off the registration years, the majority of cars being sold were registered between 1991 and 2016. 

### Identifying popular car brands advertised on eBay(Germany)
With the preliminary cleaning completed, we can start off with some basic analysis of the dataset.

We can look to the brands of cars advertised on the website to see which brands were the most popular amongst the buyers and sellers. Based on the distribution of brands we can decide if we want to move forward with all the data or exclude some.

In [21]:
popular_brands = autos["brand"].value_counts(normalize=True,dropna=False)
print(popular_brands)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

We can see that the top 5 advertised brands are all German manufacturers and account ~60% of the total number of ads crawled. This would make sense since this data is a reflection of eBay(Germany).

Based off the distribution, we can take a closer look at car manufacturers who account for atleast 1% of ads on eBay.

In [22]:
top_brands = popular_brands[popular_brands >= 0.01].index
top_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart',
       'citroen', 'toyota', 'hyundai'],
      dtype='object')

In [23]:
brand_prices = {}

for each in top_brands:
    brand_temp = autos[autos['brand'] == each]
    mean_price = brand_temp['price_$'].mean()
    mean_price = format(mean_price,".2f")
    brand_prices[each] = float(mean_price)
    
brand_prices

{'volkswagen': 5402.41,
 'bmw': 8332.82,
 'opel': 2975.24,
 'mercedes_benz': 8628.45,
 'audi': 9336.69,
 'ford': 3749.47,
 'renault': 2474.86,
 'peugeot': 3094.02,
 'fiat': 2813.75,
 'seat': 4397.23,
 'skoda': 6368.0,
 'nissan': 4743.4,
 'mazda': 4112.6,
 'smart': 3580.22,
 'citroen': 3779.14,
 'toyota': 5167.09,
 'hyundai': 5365.25}

Looking at the mean price in $ of the cars advertised by their associated brand it is easy to conclude that luxury brands like Audi, BMW and Mercedes-Benz cost more than economy brands like Volkswagen, Ford and Toyota.

Audi cars have the highest mean price, while the lowest is Renault. 

In [24]:
brand_mileage = {}

for each in top_brands:
    brand_temp = autos[autos['brand'] == each]
    mean_km = brand_temp['odometer_km'].mean()
    mean_km = format(mean_km,".2f")
    brand_mileage[each] = float(mean_km)  

brand_mileage

{'volkswagen': 128707.16,
 'bmw': 132572.51,
 'opel': 129310.04,
 'mercedes_benz': 130788.36,
 'audi': 129157.39,
 'ford': 124266.01,
 'renault': 128071.33,
 'peugeot': 127153.63,
 'fiat': 117121.97,
 'seat': 121131.3,
 'skoda': 110848.56,
 'nissan': 118331.0,
 'mazda': 124464.03,
 'smart': 99326.78,
 'citroen': 119694.19,
 'toyota': 115944.35,
 'hyundai': 106442.31}

In [25]:
top_brands_select = pd.DataFrame(pd.Series(brand_prices).sort_values(ascending=False), columns=["mean_price_$"])
top_brands_select["mean_mileage_km"] = pd.Series(brand_mileage)
top_brands_select

mean_price_$  mean_mileage_km
audi                9336.69        129157.39
mercedes_benz       8628.45        130788.36
bmw                 8332.82        132572.51
skoda               6368.00        110848.56
volkswagen          5402.41        128707.16
hyundai             5365.25        106442.31
toyota              5167.09        115944.35
nissan              4743.40        118331.00
seat                4397.23        121131.30
mazda               4112.60        124464.03
citroen             3779.14        119694.19
ford                3749.47        124266.01
smart               3580.22         99326.78
peugeot             3094.02        127153.63
opel                2975.24        129310.04
fiat                2813.75        117121.97
renault             2474.86        128071.33

Comparing the mean price of cars advertised across the brands and their mileage there does not appear to be any visible correlation between the 2 variables. 

The most and least expensive car brands (Audi and Renault) have very similar odometer readings which suggests that the mileage does not have much impact on the price of the car. Other factors such as internal aesthetics, car condition etc. could also be contributing to the difference in price.

Based off mean price and mileage, the best brand of car to buy would be 'Smart'. It has a reasonable price not too far off from the lowest while having significantly lesser tread on its tyres.

## Changing categorical string values from German to English

During our intital inspection of the dataset, we observed that some of the variables that were categorical in nature stored their responses in German. For ease of understanding to a global audience, we proceed to convert the text from German to English.

4 variables in particular, 'gearbox','vehicle_type','fuel_type' and 'unrepaired_damage' would benefit tremendously with a shift from German to English.

In [26]:
print(autos['gearbox'].value_counts(dropna=False))
print('\n')
print(autos['vehicle_type'].value_counts(dropna=False))
print('\n')
print(autos['fuel_type'].value_counts(dropna=False))
print('\n')
print(autos['unrepaired_damage'].value_counts(dropna=False))

manuell      34715
automatik     9856
NaN           2110
Name: gearbox, dtype: int64


limousine     12598
kleinwagen    10585
kombi          8930
bus            4031
cabrio         3016
NaN            2704
coupe          2462
suv            1965
andere          390
Name: vehicle_type, dtype: int64


benzin     28540
diesel     14032
NaN         3318
lpg          649
cng           71
hybrid        37
elektro       19
andere        15
Name: fuel_type, dtype: int64


nein    33834
NaN      8307
ja       4540
Name: unrepaired_damage, dtype: int64


In [27]:
gearbox = {
    'manuell': 'manual',
    'automatik': 'automatic'}

unrepaired_damage = {
    'nein': 'no',
    'ja': 'yes'}

autos['gearbox'] = autos['gearbox'].map(gearbox)
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(unrepaired_damage)
autos['fuel_type'] = autos['fuel_type'].replace(['benzin', 'elektro', 'andere'], ['gasoline', 'electric', 'other'])
autos['vehicle_type'] = autos['vehicle_type'].replace(['limousine', 'kleinwagen', 'kombi', 'cabrio', 'andere'], ['sedan', 'small car', 'stationwagon', 'convertible', 'other'])
autos['model'] = autos['model'].replace(['andere'], ['other'])

To improve readability we change the date format in the 3 date columns ('date_crawled', 'ad_created', 'last_seen') to numeric data and do away with the dashes separating the mm dd and yy.

In [28]:
dates = ['date_crawled','ad_created','last_seen']

for each in dates:
    autos[each] = (autos[each].str[:10]).str.replace('-','').astype(int)

#### Does the transmission and accident history of a used car affect it's price ?

2 columns in the dataset describe the transmission(manual/automatic) and accident history(yes/no) of the cars respectively. We can use these variables to further look at how the car prices fluctuate.

In [29]:
autos_damaged = autos[autos['unrepaired_damage'] == 'yes']
autos_not_damaged = autos[autos['unrepaired_damage'] == 'no']

mean_price_damaged = float(autos_damaged['price_$'].mean())
mean_price_damaged = format(mean_price_damaged,".2f")
mileage_damaged = float(autos_damaged['odometer_km'].mean())
mileage_damaged = format(mileage_damaged,".2f")

mean_price_not_damaged = float(autos_not_damaged['price_$'].mean())
mean_price_not_damaged = format(mean_price_not_damaged,".2f")
mileage_not_damaged = float(autos_not_damaged['odometer_km'].mean())
mileage_not_damaged = format(mileage_not_damaged,".2f")

df = {'mean_price_$':[mean_price_damaged,mean_price_not_damaged],'mean_mileage_km':[mileage_damaged,mileage_not_damaged]}
damage_comp = pd.DataFrame(df,index=['unrepaired_damage_yes','unrepaired_damage_no'])
damage_comp
    

mean_price_$ mean_mileage_km
unrepaired_damage_yes      2241.15       135356.83
unrepaired_damage_no       7164.03       122912.31

As expected cars that have some form of unrepaired damage were significantly cheaper than their pristine counterparts.

In [30]:
autos_manual = autos[autos['gearbox'] == 'manual']
autos_automatic = autos[autos['gearbox'] == 'automatic']

mean_price_manual = float(autos_manual['price_$'].mean())
mean_price_manual = format(mean_price_manual,".2f")
mileage_manual = float(autos_manual['odometer_km'].mean())
mileage_manual = format(mileage_manual,".2f")

mean_price_automatic = float(autos_automatic['price_$'].mean())
mean_price_automatic = format(mean_price_automatic,".2f")
mileage_automatic = float(autos_automatic['odometer_km'].mean())
mileage_automatic = format(mileage_automatic,".2f")

df1 = {'mean_price_$':[mean_price_manual,mean_price_automatic],'mean_mileage_km':[mileage_manual,mileage_automatic]}
gearbox_comp = pd.DataFrame(df1,index=['manual','automatic'])
gearbox_comp

mean_price_$ mean_mileage_km
manual         4716.71       126982.57
automatic     10972.72       121885.15

Automatic transmission cars are significantly more expensive than manual transmission cars. 

Let us look at the brands that offer automatic transmission to see if it is limited to luxury brands or do economy brands offer it as well. Furthermore, we will also see what the mean price and mileage is for brands that offer automatic transmission.

In [31]:
print((autos_automatic['brand'].value_counts(normalize=True)),"\n")
automatic_brands = (autos_automatic['brand'].value_counts(normalize=True)).index

automatic_cars = {}
automatic_miles = {}

for each in automatic_brands:
    brand_temp = autos_automatic[autos_automatic['brand'] == each]
    mean_price = brand_temp['price_$'].mean()
    mean_price = format(mean_price,".2f")
    mean_km = brand_temp['odometer_km'].mean()
    mean_km = format(mean_km,".2f")
    automatic_cars[each] = float(mean_price)
    automatic_miles[each] = float(mean_km)
    
temp = pd.Series(automatic_cars)

automatic_automobiles = pd.DataFrame(temp, columns=["mean_price_$"])
automatic_automobiles["mean_mileage_km"] = pd.Series(automatic_miles)
automatic_automobiles

mercedes_benz     0.251319
bmw               0.174107
audi              0.134943
volkswagen        0.131494
opel              0.051847
smart             0.049107
ford              0.026177
sonstige_autos    0.018466
renault           0.014712
volvo             0.013393
porsche           0.012480
chevrolet         0.009030
skoda             0.008929
citroen           0.008624
kia               0.008117
chrysler          0.008015
nissan            0.007711
peugeot           0.007610
jeep              0.007305
toyota            0.007204
hyundai           0.005885
jaguar            0.005479
mitsubishi        0.004972
honda             0.004870
mazda             0.004261
land_rover        0.004160
seat              0.003450
fiat              0.002841
mini              0.002334
alfa_romeo        0.002131
subaru            0.001725
suzuki            0.001725
saab              0.001725
daihatsu          0.001623
lancia            0.000812
rover             0.000710
daewoo            0.000507
l

mean_price_$  mean_mileage_km
mercedes_benz       11216.38        129301.57
bmw                 12376.05        129085.08
audi                13836.55        125135.34
volkswagen          10489.09        118499.23
opel                 3630.90        124960.86
smart                3821.94         95712.81
ford                 8614.15        114089.15
sonstige_autos      18548.65         96373.63
renault              3894.41        118896.55
volvo                6983.08        138787.88
porsche             49767.75         93495.93
chevrolet            9656.35        112247.19
skoda               11697.27         93806.82
citroen              6568.27        120294.12
kia                  8135.36        118562.50
chrysler             4762.06        132405.06
nissan               7546.92        107236.84
peugeot              4357.97        116800.00
jeep                12890.22        128750.00
toyota               7998.99         96197.18
hyundai              8217.36        115086.21
jaguar              13512.78        123796.30
mitsubishi           6518.31        123469.39
honda                5535.29        110937.50
mazda                7857.10        106547.62
land_rover          24870.90        114390.24
seat                10177.06         86911.76
fiat                 3610.04         88571.43
mini                15212.74         61739.13
alfa_romeo           5558.95        123571.43
subaru               3969.29        140000.00
suzuki               3739.94        101176.47
saab                 3970.71        135882.35
daihatsu             1316.56        104687.50
lancia               4081.25        104375.00
rover                4337.14        146428.57
daewoo               1529.80         95000.00
lada                 2850.00         90000.00
trabant               450.00          5000.00

~70% of the cars which have automatic transmission belong to 4 brands: 'Mercedes-Benz, BMW, Audi and Volkswagen'. The average price of a car from either of these brands with auto transmission is \$10000 or more. This probably is an indication that automatic transmission is considered a luxury in Germany.